src/libos/src/fs/syscalls.rs

```diff
@@ -35,6 +37,64 @@ pub fn do_eventfd2(init_val: u32, flags: i32) -> Result<isize> {
     Ok(fd as isize)
 }
 
+pub fn do_timerfd_create(clockid: clockid_t, flags: i32) -> Result<isize> {
+    debug!("timerfd: clockid {}, flags {} ", clockid, flags);
+
+    let clockid = time::ClockID::from_raw(clockid)?;
+
+    let inner_flags =
+        TimerCreationFlags::from_bits(flags).ok_or_else(|| errno!(EINVAL, "invalid flags"))?;
+    let file_ref: Arc<dyn File> = {
+        let timer = TimerFile::new(clockid, inner_flags)?;
+        Arc::new(timer)
+    };
+
+    let fd = current!().add_file(
+        file_ref,
+        inner_flags.contains(TimerCreationFlags::TFD_CLOEXEC),
+    );
+    Ok(fd as isize)
+}
+
+pub fn do_timerfd_settime(
+    fd: FileDesc,
+    flags: i32,
+    new_value_ptr: *const itimerspec_t,
+    old_value_ptr: *mut itimerspec_t,
+) -> Result<isize> {
+    from_user::check_ptr(new_value_ptr)?;
+    let inner_flags =
+        TimerSetFlags::from_bits(flags).ok_or_else(|| errno!(EINVAL, "invalid flags"))?;
+
+    let current = current!();
+    let file = current.file(fd)?;
+    let timerfile = file.as_timer()?;
+    let new_value = unsafe { new_value_ptr.read() };
+    let old_value = timerfile.set_time(inner_flags, &new_value)?;
+    if !old_value_ptr.is_null() {
+        from_user::check_mut_ptr(old_value_ptr)?;
+        unsafe {
+            old_value_ptr.write(old_value);
+        }
+    }
+    Ok(0)
+}
+
+pub fn do_timerfd_gettime(fd: FileDesc, curr_value_ptr: *mut itimerspec_t) -> Result<isize> {
+    if curr_value_ptr.is_null() {
```

Comment:
this check is useless, `check_mut_ptr` has already checked if is null

Reply:
Done